ADR/0032-decoupling-deployment.md

```diff
@@ -0,0 +1,201 @@
+# 32. Decoupling Deployment
+
+Date: 2023-11-17
+
+## Status
+
+Draft
+
+## Authors
+
+- Ralph Bean
+
+## Context
+
+Since the beginning of our project, we've had an emphasis on providing an integrated experience for
+the user, automating all steps through build, test, deployment, and release to higher environments.
+
+Some challenges:
+
+- Our controllers' APIs are highly coupled. For example, it isn't feasible to try and use
+  [integration-service] today without [application-service] and [gitops-service]. This creates
+  a high barrier to engage with some of the useful subsystems we've built. A user (and potential
+  contributor) has to adopt the whole thing, or nothing.
+- Today, we can generate simple but often incorrect deployment manifests for users' applications,
+  which we deploy via the application gitops repo. The user cannot interact with or influence these
+  resources directly; they have to do so through the Application/Component API. If we permit them to
+  directly provide deployment resources through that API, the API will become a leaky abstraction.
+  If don't permit them to provide their own resources, and instead try to handle every application
+  configuration case in the HAS generation code, we will struggle to keep up with users' demands for
+  the many and varied kinds of application deployments.
+
+See also [RHTAP-1873](https://issues.redhat.com/browse/RHTAP-1873).
+
+## Decision
+
+We are going to decouple the deployment from the build, test, and release portions of our system.
+
+**Deployment**:
+
+- The [Environment], [SnapshotEnvironmentBinding], and [GitOpsDeploymentManagedEnvironment]
+  resources will be deprecated and eventually decomissioned.
+- The [application-service] will stop generating GitOps repo content.
+- HAC will stop rendering [Environments] and their status.
+- If a user wants to make use of deployment capabilities, we will promote the usage of [renovatebot]
+  to propagate _released_ images to their self-managed gitops repo as pull requests.
+
+**Test and promotion**:
+
+- We will retain the [DeploymentTargetClaim], [DeploymentTarget], and [DeploymentTargetClass] APIs.
+  Collectively, call these the "DTC" APIs. These should have no dependence on other AppStudio APIs.
+  Only [integration-service] should depend on them. The controllers that support the DTC API should
+  be organized into their own project: "the [deployment-target-operator]", or something like that.
+- [integration-service] should be modified to create and destroy [DeploymentTargetClaims] directly
+  instead of [Environment] resources. Users will be expected to provide integration test pipelines
+  that accept a kubeconfig and snapshot as they do now, but those pipelines should now take initial
+  steps to *deploy the instance of their application to be tested* to the deployment target provided
+  by [integration-service] using the provided kubeconfig.
```

Comment:
I assume that these pipelines will not have Renovate integrated into them. The pipelines themselves will have to understand how to crack open the Snapshot to separate which components need to be deployed to an environment and which components will be the "public interface" or will drive the installation.

Reply:
We might consider either expanding the Snapshot or providing additional information that helps with this. Some things like resource details (replicas, limits etc.) and environment variables are only propagated through `SnapshotEnvironmentBindings` currently afaik.

Reply:
Ack. I think here, we're giving up control of those details entirely to the user. If they have a helm chart which deploys their app (including specifying replicas, limits, etc) then they just invoke that as one of the first tasks in their integration test pipelinerun. If they have an openshift template which deploys their app (including specifying replicas, limits, etc) then they just process and apply that as one of the first tasks in their integration test pipelinerun.

Let's keep those details _out_ of the Snapshot. It is deliberate feature reduction.

Reply:
I don't think that we need to add additional details to the snapshot. It already contains the git reference for each component, so the user can clone the matching manifest for the image with those.

An alternative would be to create another component that just contains manifests (in case the users doesn't want to keep manifests alongside the source code).